### PR TITLE
[vcpkg] Windows XP support

### PR DESCRIPTION
**Pull request by Mixaill**
## New variables in triplets
* `VCPKG_LINKER_SUBSYSTEM`
* `VCPKG_LINKER_SUBSYSTEM_MINIMUM_VERSION`

## Changes in `vcpkg_build_msbuild()`
* added support for `VCPKG_LINKER_SUBSYSTEM` and `VCPKG_LINKER_SUBSYSTEM_MINIMUM_VERSION`

## Changes in `vcpkg_configure_cmake()`
* added support for `VCPKG_PLATFORM_TOOLSET`
* added support for `VCPKG_LINKER_SUBSYSTEM` and `VCPKG_LINKER_SUBSYSTEM_MINIMUM_VERSION`
* force `msbuild` for custom toolsets

## Changes in `vcpkg_configure_meson()`
* added support for `VCPKG_LINKER_SUBSYSTEM` and `VCPKG_LINKER_SUBSYSTEM_MINIMUM_VERSION`
* abort build for custom toolsets

## Changes in `vcpkg_configure_qmake()`
* abort build for custom toolsets
* abort build for custom subsystems

## Changes in `vcpkg.exe`
* **coff_file_reader**: implement reading of `subsystem` and `subsystem_version`
* **post_build_lint**: implement `dll_subsystem_version` check

## Changes in documentation
* new triplet variables were described

## Remarks and known limitations
* `meson` and `qmake` does not support custom toolsets so build will be simply aborted.

* `qmake` does not support(?) to set linker flags through command line or environment variables so build will be simply aborted.

* `VCPKG_LINKER_SUBSYSTEM` will be automatically set to `Console` if 
   * `VCPKG_LINKER_SUBSYSTEM` is not set
   * `VCPKG_LINKER_SUBSYSTEM_MINIMUM_VERSION` is set

## Checklist
* build
    * `subsystem`/`subsystem_min_version`
       - [x] cmake
       - [x] meson
       - [x] msbuild
       - [x] qmake (abort build)
    * `toolset`
       - [x] cmake (force `msbuild` if custom and not xp)
       - [x] meson (abort build if custom and not xp)
       - [x] msbuild
       - [x] qmake  (abort build if custom)
* post-build validation
   - [x] check PE minimum subsystem version
* doc
   - [x] describe new triplet file variables

### CODE REVIEW

**Comment by Mixaill** [comment]
Meson and qmake support do not support `VCPKG_PLATFORM_TOOLSET` at all?

**Comment by codicodi** [comment]
Meson picks up compiler from the environment set by vcpkg, similarly to how CMake+Ninja works. AFAIK buildsystems other than MSBuild don't have a notion of toolset.

**Comment by Mixaill** [comment]
Looks like *_xp toolsets doing these tasks:
* set subsystem minimum version to 5.02
* switch Windows SDK to 7.1
* define `_USING_V110_SDK71_`
* set `/Zc:threadSafeInit-` for ATL dynamic libraries

**Comment by Mixaill** [comment]
>AFAIK buildsystems other than MSBuild don't have a notion of toolset.

Yes, looks like toolsets is a MSBuild-only stuff.

Maybe we should add prefix `msbuild_` to all places when we use toolsets? @alexkaratarakis

**Comment by alroyso** [comment]
vs2017 报错 
CMake Error at CMakeLists.txt:47 (project):
  Generator

    Ninja

  does not support toolset specification, but toolset

    v141_xp

  was specified.


CMake Error: CMAKE_C_COMPILER not set, after EnableLanguage

**Comment by Mixaill** [comment]
Ready for testing

File: docs/users/triplets.md [hunk 1]
```diff
@@ -35,6 +45,11 @@ Specifies the C/C++ compiler toolchain to use.
 
 This can be set to `v141`, `v140`, or left blank. If left blank, we select the latest compiler toolset available on your machine.
 
+### VCPKG_PLATFORM_TOOLSET_SUFFIX
```
Comment (ras0219-msft) [inline docs/users/triplets.md]: It really feels like this shouldn't need a separate variable -- can't we just set `VCPKG_PLATFORM_TOOLSET` to `140_xp`?

File: scripts/cmake/vcpkg_build_msbuild.cmake [hunk 2]
```diff
@@ -35,6 +35,12 @@
 ## ### TARGET
 ## The MSBuild target to build. (``/t:<TARGET>``)
 ##
+## ### SUBSYSTEM
```
Comment (ras0219-msft) [inline scripts/cmake/vcpkg_build_msbuild.cmake]: I think that we should avoid adding these options for now; I don't really see where it ever makes sense to override them on a per-port basis. We should always use the `VCPKG_LINKER_SUBSYSTEM` setting.

File: scripts/cmake/vcpkg_configure_cmake.cmake [hunk 3]
```diff
@@ -171,10 +184,28 @@ function(vcpkg_configure_cmake)
...
+        if(DEFINED VCPKG_LINKER_SUBSYSTEM AND NOT DEFINED VCPKG_LINKER_SUBSYSTEM_MIN_VERSION)
+            list(APPEND VCPKG_LINKER_FLAGS "/SUBSYSTEM:${VCPKG_LINKER_SUBSYSTEM}")
+        endif()
+
         list(APPEND _csc_OPTIONS_RELEASE
             "-DCMAKE_SHARED_LINKER_FLAGS_RELEASE=/DEBUG /INCREMENTAL:NO /OPT:REF /OPT:ICF ${VCPKG_LINKER_FLAGS}"
             "-DCMAKE_EXE_LINKER_FLAGS_RELEASE=/DEBUG /INCREMENTAL:NO /OPT:REF /OPT:ICF ${VCPKG_LINKER_FLAGS}"
         )
+
+        list(APPEND _csc_OPTIONS_DEBUG
```
Comment (ras0219-msft) [inline scripts/cmake/vcpkg_configure_cmake.cmake]: Ouch, this was an oversight. Thanks for fixing it!

File: scripts/cmake/vcpkg_configure_qmake.cmake [hunk 4]
```diff
@@ -21,24 +21,32 @@
...
 function(vcpkg_configure_qmake)
-    cmake_parse_arguments(_csc "" "SOURCE_PATH" "OPTIONS" ${ARGN})
-    
-    # Find qmake exectuable 
-    find_program(QMAKE_COMMAND NAMES qmake.exe PATHS ${CURRENT_INSTALLED_DIR}/tools/qt5)
-    
-    if(NOT QMAKE_COMMAND)
-        message(FATAL_ERROR "vcpkg_configure_qmake: unable to find qmake.")
-    endif()
+cmake_parse_arguments(_csc "" "SOURCE_PATH" "OPTIONS" ${ARGN})
```
Comment (ras0219-msft) [inline scripts/cmake/vcpkg_configure_qmake.cmake]: Indentation

File: triplets/x64-windows-static-xp.cmake [hunk 5]
```diff
@@ -0,0 +1,6 @@
+set(VCPKG_TARGET_ARCHITECTURE x64)
```
Comment (ras0219-msft) [inline triplets/x64-windows-static-xp.cmake]: We do not intend to officially test xp, so we should not add these triplets by default.

File: toolsrc/include/vcpkg/build.h [hunk 6]
```diff
@@ -94,9 +94,12 @@ namespace vcpkg::Build
         std::string cmake_system_name;
         std::string cmake_system_version;
         Optional<std::string> platform_toolset;
+        Optional<std::string> platform_toolset_suffix;
```
Comment (ras0219-msft) [inline toolsrc/include/vcpkg/build.h]: I don't see this ever consumed inside the sources; if it's not needed let's not bother exfiltrating it.

File: toolsrc/include/vcpkg/build.h [hunk 7]
```diff
@@ -94,9 +94,12 @@ namespace vcpkg::Build
...
         Optional<fs::path> visual_studio_path;
-        Optional<std::string> external_toolchain_file;
+        Optional<std::string> external_toolchain_file;
```
Comment (ras0219-msft) [inline toolsrc/include/vcpkg/build.h]: trailing whitespace

**Review by ras0219-msft — CHANGES_REQUESTED** [review]

File: docs/users/triplets.md (hunk 1 above)
Comment (Mixaill) [reply docs/users/triplets.md]: merged into one variable

File: scripts/cmake/vcpkg_build_msbuild.cmake (hunk 2 above)
Comment (Mixaill) [reply scripts/cmake/vcpkg_build_msbuild.cmake]: removed

File: scripts/cmake/vcpkg_configure_qmake.cmake (hunk 4 above)
Comment (Mixaill) [reply scripts/cmake/vcpkg_configure_qmake.cmake]: fixed

File: toolsrc/include/vcpkg/build.h (hunk 6 above)
Comment (Mixaill) [reply toolsrc/include/vcpkg/build.h]: fixed

File: toolsrc/include/vcpkg/build.h (hunk 7 above)
Comment (Mixaill) [reply toolsrc/include/vcpkg/build.h]: fixed

File: triplets/x64-windows-static-xp.cmake (hunk 5 above)
Comment (Mixaill) [reply triplets/x64-windows-static-xp.cmake]: removed

**Comment by Mixaill** [comment]
@ras0219-msft fixed

**Review by ras0219-msft — APPROVED** [review]

File: scripts/cmake/vcpkg_configure_cmake.cmake [hunk 8]
```diff
@@ -57,53 +57,34 @@ function(vcpkg_configure_cmake)
...
-        AND NOT VCPKG_CMAKE_SYSTEM_NAME STREQUAL "WindowsStore" 
-        AND VCPKG_PLATFORM_TOOLSET MATCHES "^v[0-9][0-9][0-9]?(_xp)?$"
-    )
-        set(USE_NINJA True)
-    endif()
-
-    #select preferable generator
     if(_csc_GENERATOR)
         set(GENERATOR ${_csc_GENERATOR})
+    elseif(_csc_PREFER_NINJA AND NOT VCPKG_CMAKE_SYSTEM_NAME STREQUAL "WindowsStore" AND NOT _csc_HOST_ARCHITECTURE STREQUAL "x86")
```
Comment (Mixaill) [inline scripts/cmake/vcpkg_configure_cmake.cmake]: It will break custom toolsets like `LLVM-vs2014`. 

There is need to check for `VCPKG_PLATFORM_TOOLSET MATCHES "^v[0-9][0-9][0-9]?(_xp)?$"`

**Comment by VikingExplorer** [comment]
So, after reading all that & rechecking the triplet documentation, I'm still not sure what change I have to make in order to support Windows XP.  

Is it:  set(VCPKG_PLATFORM_TOOLSET v141_xp)

**Comment by Mixaill** [comment]
```
set(VCPKG_LINKER_SUBSYSTEM Console)
set(VCPKG_LINKER_SUBSYSTEM_MIN_VERSION 5.01)
set(VCPKG_PLATFORM_TOOLSET v141_xp)
```

**Comment by simonowen** [comment]
Should this already work using just the settings above in a new triplet?

I found I also had to modify toolsrc\src\vcpkg\vcpkgpaths.cpp to accept the _xp suffix when matching toolset versions, otherwise it fails with `"Could not find Visual Studio instance with v141_xp toolset."`.  That also seems to be what @VikingExplorer reported in issue #2580.

**Comment by Mixaill** [comment]
Should works for `v141_xp` now.

It would be nice to fill available toolsets list from MSBuild directories:
* `<VS 2017 dir>/Common7/IDE/VC/VCTargets/Platforms/*/PlatformToolsets/*`
* `<Program Files x86 dir>/MSBuild/Microsoft.Cpp/v4.0/*/Platforms/*/PlatformToolsets/*`

It will enable LLVM/Clang and Clang/C2. What do you think, @ras0219-msft ?

**Comment by past-due** [comment]
@Mixaill: What remaining pieces are needed to properly support `v140_xp`? Am happy to dig in and contribute (it's looking like I'll need this functionality fairly soon).

**Comment by Stanzilla** [comment]
Why even support WinXP?

**Comment by VikingExplorer** [comment]
As of 2017, more xp users than Mac os Sierra and linux combined.

**Comment by Stanzilla** [comment]
Adding support for it to modern systems is not going to help make that number go down though.

**Comment by VikingExplorer** [comment]
Odd emotional bias that businesses don't usually have. We want to supply
what the market demands. Apparently, Microsoft thinks the same, providing
_xp toolchains for both v140 and v141.

**Comment by MikeGitb** [comment]
Actually, I don't think @stanzilla's comment is emotional. The longer you (or someone else) supports a system, the longer it will be used by your customers. 

That being said, I don't think supporting windows. XP is a bad thing as long as it doesn't eat up lots of resources and/or inhibits new feature development. 

In the end this is an OS project. If someone wants to spend the time to add xp support and others are not expected to do the same - why not?

**Comment by VikingExplorer** [comment]
> The longer you (or someone else) supports a system, the longer it will be used by your customers.

1) this is reversing cause and effect.
2) it implies an agenda, a bias.

Look at all the energy & excitement about supporting Linux, yet Linux is at 1.6% & XP is at 6.9% market share. ref: [Operating System Share by Version](https://www.netmarketshare.com/operating-system-market-share.aspx?options=%7B%22filter%22%3A%7B%22%24and%22%3A%5B%7B%22deviceType%22%3A%7B%22%24in%22%3A%5B%22Desktop%2Flaptop%22%5D%7D%7D%5D%7D%2C%22dateLabel%22%3A%22Trend%22%2C%22attributes%22%3A%22share%22%2C%22group%22%3A%22platformVersion%22%2C%22sort%22%3A%7B%22share%22%3A-1%7D%2C%22id%22%3A%22platformsDesktopVersions%22%2C%22dateInterval%22%3A%22Monthly%22%2C%22dateStart%22%3A%222017-03%22%2C%22dateEnd%22%3A%222018-02%22%2C%22segments%22%3A%22-1000%22%7D)

This is like coming into a donut shop, eating a big mac, and saying "the longer you keep selling donuts, the longer we're going to eat them"...

**Comment by VikingExplorer** [comment]
It seems it's not ready yet:

`set(VCPKG_TARGET_ARCHITECTURE x86)
set(VCPKG_CRT_LINKAGE dynamic)
set(VCPKG_LIBRARY_LINKAGE static)

set(VCPKG_LINKER_SUBSYSTEM Console)
set(VCPKG_LINKER_SUBSYSTEM_MIN_VERSION 5.01)
set(VCPKG_PLATFORM_TOOLSET v141_xp)
`
Result:
Installing package boost-beast[core]:x64-windows-static-md... done
Elapsed time for package boost-beast:x64-windows-static-md: 4.482 s
Starting package 59/116: boost-modular-build-helper:x86-windows-static-md
Building package boost-modular-build-helper[core]:x86-windows-static-md...
**Could not find Visual Studio instance with v141_xp toolset.**

**Comment by ryanfitzsimon** [comment]
Is this still being worked on? There don't seem to have been any changes for a few months and Mixaill/triplet-toolset now conflicts with master.

**Comment by ras0219-msft** [comment]
We do want to get this merged at some point, however xp is lower priority than many of the other outstanding PRs.

**Comment by Mixaill** [comment]
I will fix this in the near future.

**Comment by VikingExplorer** [comment]
Thank you @Mixaill 

As of Mar 2017, XP was 2.76 times more popular than Linux.

**Comment by MikeGitb** [comment]
@VikingExplorer: I've no Idea, what the real numbers are (might be better, might be worse) but I'd claim that the market share of different Desktop-OSes (likely measured by some webpage traffic) is only very loosely related to the question of what OS is most commonly targeted by current c++ developers. E.g. it ignores all the server systems and the whole embedded market and not every user out there is equally likely to use my newly developed c++ program.

And I don't think I'm reversing cause and effect when I'm saying customers are more likely to upgrade their OS if more and more of their favorite software isn't running on the old one. That isn't too say, that support for old OS can be an important factor of why your software is bought in the first place (especially depending on what your competitors do).

However, I believe the software industry as a whole is doing itself a disservice by spending time and resources on the continued support of windows XP (at least in classic desktop environments).

**Comment by VikingExplorer** [comment]
@MikeGitb I don't give a shit about your opinion about the fictional "software industry as a whole" or your subjective ideas about which OS's are cool & which are not.  It's business, that's all.  MS supports v140_xp & v141_xp. Period.  End of useless discussion.

**Comment by ras0219-msft** [comment]
To conclude the discussion about whether XP should or should not be supported by vcpkg:

- We will accept core tool support
- We will not be regularly building the package graph on XP
  - Therefore, we will not have built-in triplets for XP
- We will not block package updates based on whether they drop support for XP
- We will accept PRs fixing XP support if they do not introduce a significant additional maintenance burden

We understand XP support is a controversial topic for many, however please keep discussions factual, civil, and not targetted at individuals.

**Comment by jrampon** [comment]
Hello - what's the plan with this PR? Can we expect to see it merged into master sometime soon? How can I help? @alexkaratarakis @ras0219

**Comment by heydojo** [comment]
@VikingExplorer 

> As of 2017, more xp users than Mac os Sierra and linux combined.

It's now 2019 and without a source for your claim, that's exactly all it is : a claim.

The first link on [DDG](https://duckduckgo.com/) and the picture is very different now.
[https://www.netmarketshare.com/operating-system-market-share.aspx](https://www.netmarketshare.com/operating-system-market-share.aspx?options=%7B%22filter%22%3A%7B%22%24and%22%3A%5B%7B%22deviceType%22%3A%7B%22%24in%22%3A%5B%22Desktop%2Flaptop%22%5D%7D%7D%5D%7D%2C%22dateLabel%22%3A%22Trend%22%2C%22attributes%22%3A%22share%22%2C%22group%22%3A%22platformVersion%22%2C%22sort%22%3A%7B%22share%22%3A-1%7D%2C%22id%22%3A%22platformsDesktopVersions%22%2C%22dateInterval%22%3A%22Monthly%22%2C%22dateStart%22%3A%222018-07%22%2C%22dateEnd%22%3A%222019-06%22%2C%22segments%22%3A%22-1000%22%7D)

> Extended support for Windows XP ended on April 8, 2014

> As of July 2019, 1.71% of Windows PCs run Windows XP.

https://en.wikipedia.org/wiki/Windows_XP

Considering that vcpkg requires WAN access to be able download files, a Computer running XP in the wild connected to the Internet is the kind of crazy which does not have any place in today's modern world of interconnected devices.

> Odd emotional bias that businesses don't usually have. We want to supply
what the market demands. Apparently, Microsoft thinks the same, providing
_xp toolchains for both v140 and v141.

Firstly, businesses may still use XP embedded. Not XP. Although even XP embedded is becoming more and more rare. And XP embedded machines such as those used in terminals are only LAN facing and largely never join a WAN.

It's interesting that you are attempting to derail the discussion by attempting to claim emotional investment in a topic revolving around a depreciated and unsupported _Computer Operating system_. And as ridiculous as it may seem, you have also unfortunately not evaluated your own bias before commenting.
By asserting without basis or value that the market wants something simply because _you_ want something, you are attempting to create an availability cascade, are partaking in a base rate fallacy, you have a bias blind spot, an empathy gap and your default position by advocating Windows XP as a target for vcpkg is one of hyperbolically discounting.

In short, please extend some courtesy to others who want to discuss a topic with you and please learn that just because someone has a factual basis for fundamentally disagreeing with you, that in doing so, that does not simply make them _wrong_.
Thanks.

**Comment by VikingExplorer** [comment]
@heydojo Any objective analysis of our responses would conclude that your comments are emotional, not factual.  I go strictly by the numbers.

1) June 2019 market share from the link:  Windows XP (3.18%) over 2 times that of Linux (1.52%).  If we support Linux, we should support XP.  If not XP, drop Linux.

2) With C++, building & deployment are two separate things. Users may be have XP or XP embedded, but regardless, they aren't building software, so WAN access is totally irrelevant.

3) Lack of XP support reflects very poorly on vcpkg. Comparing Win10 vs XP vcxproj, the differences are very small:

    <WindowsTargetPlatformVersion>10.0.17763.0</WindowsTargetPlatformVersion>
vs
    <WindowsTargetPlatformVersion>7.0</WindowsTargetPlatformVersion>

AND

    <PlatformToolset>v141</PlatformToolset>
vs
    <PlatformToolset>v141_xp</PlatformToolset>

This should be quite simple for any well designed piece of software.

**Comment by degski** [comment]
@VikingExplorer Going 'strictly by the numbers' is not a very good argument here, I think. It's a known fact, witness are the many `WannaCry` and other rubbish being distributed, that many, if not the majority of those XP installs are actually within government or government-related entities [they have the security issues]. This is true world-wide. The amount of development going on [and for] those machines is less than zero. Then there are amny many nix servers, that don't show up in the numbers. Many people actually develop software for exactly those [invisible] machines, very actively. Although there is `PlatformToolset: v141_xp`, there is no such thing as `PlatformToolset: v142_xp`, in other words it's dead. You could get by with that in the past, but ternary annual updates of the std, you're surely bound to run out of road quite quickly. I agree with others here that people should be dissuaded to use and develop for XP, Vista and very soon W7.

**Comment by VikingExplorer** [comment]
@degski I'm not expressing any opinion on whether it should be used or not. I would call that an "agenda". The plain fact is that v141_xp DOES exist and it should have been quite simple to support that in vcpkg.

**Comment by degski** [comment]
@VikingExplorer But v141 has been superseded by v142, without the XP. I am expressing the opinion that it should not be used. As ever on Windows, I will need to run the now outdated and less secure crt, iff I would like to run the software built with it. The problem is obviously not v141 vs v142, but I use software, I really like, and know well, but it uses the VS2008 run-time, that's what happens, people are lazy, companies save money, but then all of a sudden `WannaCry` takes over and your f'ed, then you wished you would have been running the latest stuff.

**Comment by VikingExplorer** [comment]
@degski I understand your opinion, but I could care less. You're not arguing with me, b/c I have no opinion on the matter.  

Your statement about v142 vs v141 is as perplexing as it is irrelevant. You seem to not understand that vendors create software for customers. If those customers want a specific version, that's up to them. Microsoft has made this decision already & provides support for v141_xp.  It makes no sense that a different part of MS has chosen not to support the capabilities of v141, either deliberately or by incompetence.

**Comment by cbezault** [comment]
The conversation here is digressing from the original point of this PR.
As @ras0219-msft has already stated:
* We will accept core tool support
* We will not be regularly building the package graph on XP
  * Therefore, we will not have built-in triplets for XP
* We will not block package updates based on whether they drop support for XP
* We will accept PRs fixing XP support if they do not introduce a significant additional maintenance burden

None of these statements have anything to do with whether developing for XP is good or bad.
None of these statements have anything to do with numbers of users of different OSs.

A decision was made concerning the set of OSs we would be regularly testing and that is it.

We've been putting in a fair amount of effort recently to improve the customization options of vcpkg and have always had the custom use-case in mind when making changes to vcpkg so it shouldn't be too hard to set up an XP vcpkg.
We will, however, bring changes into mainline vcpkg for XP support that follow the guidelines set above.

**Comment by dan-shaw** [comment]
@Mixaill Going to close this PR as this PR has been open too long. If you are interested in working on this, feel free to reopen, but it might make more sense to make a new PR (otherwise we might lose track of it).

**Comment by BillyONeal** [comment]
We already had a signoff from Robert here a long time ago meaning this is probably OK and this got lost in the shuffle; unfortunately there is nontrivial work required to merge it with current sources. @Mixaill Are you interested?

**Comment by BillyONeal** [comment]
See also https://github.com/microsoft/vcpkg/issues/1015

Thank you @Mixaill for your contributions, it's been a while so I've unlocked the PR.

If others resume personal attacks here I'll be forced to relock this and lock the issue.

**Comment by Mixaill** [comment]
@BillyONeal as I mentioned before, I'll try to adopt it during the next weeks.

**Comment by BillyONeal** [comment]
I'm reclosing this due to inactivity. Please open a new PR with changes merged with master if you would like to continue.

Note that Visual Studio is unable to service the foo_xp redists ever again due to the SHA-1 signature problem mentioned in the other issue.
